File: sol/641/641.cpp
struct MyCircularDeque {
    int k, l;
    vector<int> data;
    int begin;
    int end;


    MyCircularDeque(int k): 
        k(k), l(k + 1),
        data(k + 1, -1),
        begin(0), end(0) {}
    
    inline bool insertFront(int value) {
        if (this->isFull()){
            return false;
        }
        this->begin = (this->begin + this->k) % this->l;
        this->data[this->begin] = value;
        return true;
    }
    
    inline bool insertLast(int value) {
        if (this->isFull()){
            return false;
        }
        this->data[this->end] = value;
        this->end = (this->end + 1) % this->l;
        return true;
    }
    
    inline bool deleteFront() {
        if (this->isEmpty()){
            return false;
        }
        this->begin = (this->begin + 1) % this->l;
        return true;
    }
    
    inline bool deleteLast() {
        if (this->isEmpty()){
            return false;
        }
        this->end = (this->end + this->k) % this->l;
        return true;
    }
    
    inline int getFront() const {
        return this->isEmpty() ? -1 : this->data[this->begin];
    }
    
    inline int getRear() const {
        return this->isEmpty() ? -1 : this->data[(this->end + this->k) % this->l];
    }
    
    inline bool isEmpty() const {
        return this->begin == this->end;
    }
    
    inline bool isFull() const {
        return (this->end - this->begin + this->l) % this->l == this->k;
    }
};

File: sol/641/641.py
class MyCircularDeque:
    def __init__(self, k: int):
        self.k = k
        self.data = [-1] * (k + 1)
        self.begin = 0
        self.end = 0

    def insertFront(self, value: int) -> bool:
        if self.isFull():
            return False
        self.begin = (self.begin + self.k) % (self.k + 1)
        self.data[self.begin] = value
        return True

    def insertLast(self, value: int) -> bool:
        if self.isFull():
            return False
        self.data[self.end] = value
        self.end = (self.end + 1) % (self.k + 1)
        return True

    def deleteFront(self) -> bool:
        if self.isEmpty():
            return False
        self.begin = (self.begin + 1) % (self.k + 1)
        return True

    def deleteLast(self) -> bool:
        if self.isEmpty():
            return False
        self.end = (self.end + self.k) % (self.k + 1)
        return True

    def getFront(self) -> int:
        return -1 if self.isEmpty() else self.data[self.begin]

    def getRear(self) -> int:
        return -1 if self.isEmpty() else self.data[(self.end + self.k) % (self.k + 1)]

    def isEmpty(self) -> bool:
        return self.begin == self.end

    def isFull(self) -> bool:
        return (self.end - self.begin + self.k + 1) % (self.k + 1) == self.k

